File: src/localization_example.cpp
#include "phi_p3dx_localization/localization_node.hpp"
#include <rclcpp/rclcpp.hpp>
#include <cmath>

/**
 * @class LocalizationExample
 * @brief Exemplo didático completo de uso de LocalizationNode.
 * 
 * Demonstra como estender a classe base para implementar comportamento customizado.
 * Neste exemplo, as partículas rotacionam continuamente, permitindo observar:
 * - Como a pose estimada acompanha a rotação
 * - Como a covariância se comporta com movimento
 * - Como sobrescrever update_particles() para adicionar lógica
 * 
 * **Parâmetros:**
 * - Herda todos os parâmetros de LocalizationNode
 * - `rotation_speed` (double): Velocidade de rotação em rad/s (default: 0.5)
 * 
 * **Como usar:**
 * ```bash
 * ros2 run phi_p3dx_localization localization_node
 * ```
 */
class LocalizationExample : public LocalizationNode
{
public:
  /**
   * @brief Construtor do exemplo de localização.
   * 
   * @param node_name Nome do nó no ROS 2 (default: "localization_example")
   */
  explicit LocalizationExample(const std::string &node_name = "localization_example");
  virtual ~LocalizationExample() = default;

protected:
  /**
   * @brief Atualiza as partículas rotacionando-as continuamente.
   * 
   * Exemplo de como estender a classe base com comportamento customizado.
   * Cada partícula tem sua orientação (theta) incrementada de acordo com
   * a velocidade de rotação.
   */
  void update_particles() override;

private:
  double rotation_speed_;       ///< Velocidade de rotação em rad/s
  rclcpp::Time last_update_time_;
};

// ============ Implementação ============

LocalizationExample::LocalizationExample(const std::string &node_name)
  : LocalizationNode(node_name),
    rotation_speed_(0.5),
    last_update_time_(this->now())
{
  // Declarar e obter parâmetro de velocidade de rotação
  this->declare_parameter<double>("rotation_speed", 0.5);
  rotation_speed_ = this->get_parameter("rotation_speed").as_double();

  RCLCPP_INFO(this->get_logger(), 
    "[%s] Exemplo de localização com rotação inicializado.",
    node_name.c_str());
  RCLCPP_INFO(this->get_logger(), 
    "Velocidade de rotação: %.2f rad/s", rotation_speed_);
}

void LocalizationExample::update_particles()
{
  // Calcula delta de tempo desde a última atualização
  rclcpp::Time now = this->now();
  double dt = (now - last_update_time_).seconds();
  last_update_time_ = now;

  // Incremento de ângulo
  double delta_theta = rotation_speed_ * dt;

  // Atualiza todas as partículas
  for (auto &p : particles_) {
    p.theta += delta_theta;
    
    // Normaliza theta para [-π, π]
    while (p.theta > M_PI) {
      p.theta -= 2.0 * M_PI;
    }
    while (p.theta < -M_PI) {
      p.theta += 2.0 * M_PI;
    }
  }
}

// ============ Main ============

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<LocalizationExample>("localization_example");
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: scripts/localization_example.py
#!/usr/bin/env python3
"""
@brief Exemplo de uso de LocalizationNode em Python.

Demonstra como estender a classe base para implementar comportamento customizado.
Neste exemplo, as partículas rotacionam continuamente, permitindo observar:
- Como a pose estimada acompanha a rotação
- Como a covariância se comporta com movimento
- Exemplo de como sobrescrever update_particles() para adicionar lógica

**Parâmetros adicionais:**
- `rotation_speed` (float, default: 0.5 rad/s)

**Como usar:**
```bash
ros2 run phi_p3dx_localization localization_example_py
```
"""

import rclpy
import math
from phi_p3dx_localization.main import LocalizationNode


class LocalizationExample(LocalizationNode):
    """Exemplo de localização com rotação contínua das partículas."""

    def __init__(self, node_name='localization_example_py'):
        super().__init__(node_name)

        # Parâmetro adicional
        self.declare_parameter('rotation_speed', 0.5)
        self.rotation_speed = self.get_parameter('rotation_speed').get_parameter_value().double_value

        # Estado para controle de tempo
        self.last_update_time = self.get_clock().now()

        self.get_logger().info(
            f'[{node_name}] Exemplo de localização com rotação inicializado.'
        )
        self.get_logger().info(f'Velocidade de rotação: {self.rotation_speed:.2f} rad/s')

    def update_particles(self):
        """Sobrescreve o método de atualização para rotacionar as partículas."""
        # Calcula delta de tempo desde a última atualização
        now = self.get_clock().now()
        dt = (now - self.last_update_time).nanoseconds / 1e9  # Converte para segundos
        self.last_update_time = now

        # Incremento de ângulo
        delta_theta = self.rotation_speed * dt

        # Rotaciona todas as partículas
        for p in self.particles:
            p.theta += delta_theta

            # Normaliza theta para [-π, π]
            while p.theta > math.pi:
                p.theta -= 2.0 * math.pi
            while p.theta < -math.pi:
                p.theta += 2.0 * math.pi


def main(args=None):
    """Função principal."""
    rclpy.init(args=args)
    node = LocalizationExample()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()